x-pack/plugins/fleet/server/services/epm/kibana/assets/tag_assets.ts

```diff
@@ -31,40 +76,61 @@ interface TagAssetsParams {
   pkgName: string;
   spaceId: string;
   importedAssets: SavedObjectsImportSuccess[];
+  assetTags?: PackageSpecTags[];
 }
 
 export async function tagKibanaAssets(opts: TagAssetsParams) {
   const { savedObjectTagAssignmentService, kibanaAssets, importedAssets } = opts;
+
   const getNewId = (assetId: string) =>
     importedAssets.find((imported) => imported.id === assetId)?.destinationId ?? assetId;
   const taggableAssets = getTaggableAssets(kibanaAssets).map((asset) => ({
     ...asset,
     id: getNewId(asset.id),
   }));
+  if (taggableAssets.length > 0) {
+    const [managedTagId, packageTagId] = await Promise.all([
+      ensureManagedTag(opts),
+      ensurePackageTag(opts),
+    ]);
+    try {
+      await savedObjectTagAssignmentService.updateTagAssignments({
+        tags: [managedTagId, packageTagId],
+        assign: taggableAssets,
+        unassign: [],
+        refresh: false,
+      });
+    } catch (error) {
+      if (error.status === 404) {
+        appContextService.getLogger().warn(error.message);
+        return;
+      }
+      throw error;
+    }
 
-  // no assets to tag
-  if (taggableAssets.length === 0) {
-    return;
-  }
+    const packageSpecAssets = await getPackageSpecTags(taggableAssets, opts);
+    const groupedAssets = groupByAssetId(packageSpecAssets);
```

Comment:
There is a lot of data manipulation here, trying to explain why. 
I was previously trying to update call `updateTagAssignments` iterating directly on `packageSpecAssets`, calling by `tagId`, but for some reason the assets didn't get applied correctly. I suspect that calling the same assets on different ids asynchronously was overriding some of the assets. 

So I decided to group by `assetId` instead, so every asset gets assigned to the tag only once.